### PR TITLE
subtlecrypto: Replace `NormalizedAlgorithm` with specialized variants

### PR DESCRIPTION
**Pull request by simonwuelker**
This is a refactoring that requires some understanding of the [WebCryptoAPI spec](https://w3c.github.io/webcrypto), so I'll write slightly more than usual.

## Context
All functions in [`subtlecrypto`](https://w3c.github.io/webcrypto/#subtlecrypto-interface) receive either an object or a string to identify an algorithm. They then call [`normalize_algorithm`](https://w3c.github.io/webcrypto/#algorithm-normalization-normalize-an-algorithm) and provide a operation (`"digest"`, `"encrypt"` etc) to turn it into a `NormalizedAlgorithm`. This could for example be `Sha1` for hashing, or [`RsaHashedKeyGenParams`](https://w3c.github.io/webcrypto/#RsaHashedKeyGenParams-dictionary) for generating RSA keys. There are a lot of different `NormalizedAlgorithm` variants. [`RSASSA-PKCS1-v1_5`](https://w3c.github.io/webcrypto/#rsassa-pkcs1) alone defines four different ones.

Due to this, we currently have an ever-growing [`NormalizedAlgorithm`](https://github.com/servo/servo/blob/main/components/script/dom/subtlecrypto.rs#L722) enum that contains every algorithm in existence.

This is, of course, overly complex. When the spec creates a algorithm with operation `"generateKey"` then this algorithm will only ever generate keys, never encrypt values, and it's never going to be a value that identifies a hash.

## Proposed Changes
Instead of implementing one `normalize_algorithm(algorithm, operation)` function that returns a `NormalizedAlgorithm`, we should implement dedicated functions for each operation that return dedicated data structures.

For example, there is now `normalize_algorithm_for_digest(algorithm)`, which returns `DigestAlgorithm`. `DigestAlgorithm` is less clunky to use than the old `NormalizedAlgorithm`, because you don't have to deal with the (impossible) case of it containing AES encryption parameters for example.

The obvious downside is that this implementation deviates from the spec. However, the deviation is self-contained (everything outside of `normalize_algorithm` remains the same) and our current implementation of `normalize_algorithm` already doesn't really follow the spec.

Some tests are fixed because apparently `"get key length"` was broken before.

cc @msub2 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] There are tests for these changes

### CODE REVIEW

**Comment by msub2** [comment]
I'm supportive of this change, I was also having some thoughts in the back of my mind about things getting unwieldy with NormalizedAlgorithm, I just hadn't been in a spot to address it yet.

File: components/script/dom/subtlecrypto.rs [hunk 1]
```diff
@@ -1019,6 +1023,34 @@ fn normalize_algorithm_for_encrypt_or_decrypt(
...
+    };
+
+    rooted!(in(*cx) let value = ObjectValue(unsafe { *obj.get_unsafe() }));
+    let Ok(ConversionResult::Success(algorithm)) = Algorithm::new(cx, value.handle()) else {
+        return Err(Error::Syntax);
+    };
+
+    let name = algorithm.name.str();
+    let normalized_algorithm =
+        if name.eq_ignore_ascii_case(ALG_AES_CBC) | name.eq_ignore_ascii_case(ALG_AES_CTR) {
```
Comment (jdm) [inline components/script/dom/subtlecrypto.rs:1021]: | -> ||

**Review by jdm — APPROVED** [review]
This definitely feels worth deviating from the spec steps to make the code clearer.

File: components/script/dom/subtlecrypto.rs (hunk 1 above)
Comment (simonwuelker) [reply components/script/dom/subtlecrypto.rs]: Good catch

**Comment by simonwuelker** [comment]
That's odd, the failing tests use `subtle.digest;, but the `WebCryptoAPI/digest` tests all pass. I'll investigate...

**Comment by simonwuelker** [comment]
Apologies, I didn't do a full try run because i figured any changes should manifest in `WebCryptoAPI`. I've updated the test expectations and started https://github.com/simonwuelker/servo/actions/runs/11757527711, just to be sure.

**Comment by jdm** [comment]
> Apologies, I didn't do a full try run because i figured any changes should manifest in `WebCryptoAPI`. I've updated the test expectations and started https://github.com/simonwuelker/servo/actions/runs/11757527711, just to be sure.

This is a completely reasonable choice to me! I was surprised to see tests using the subtle interface outside of the WebCrypto tests.

**Pull request merged**